helper/schema/resource.go

```diff
@@ -56,13 +56,25 @@ var ReservedResourceFields = []string{
 // being implemented.
 type Resource struct {
 	// Schema is the structure and type information for this component. This
-	// field is required for all Resource concepts.
+	// field, or SchemaFunc, is required for all Resource concepts. To prevent
+	// storing all schema information in memory for the lifecycle of a provider,
+	// use SchemaFunc instead.
 	//
 	// The keys of this map are the names used in a practitioner configuration,
 	// such as the attribute or block name. The values describe the structure
 	// and type information of that attribute or block.
 	Schema map[string]*Schema
 
+	// SchemaFunc is the structure and type information for this component. This
+	// field, or Schema, is required for all Resource concepts. Use this field
+	// instead of Schema on top level Resource declarations to prevent storing
+	// all schema information in memory for the lifecycle of a provider.
+	//
+	// The keys of this map are the names used in a practitioner configuration,
+	// such as the attribute or block name. The values describe the structure
+	// and type information of that attribute or block.
+	SchemaFunc func() map[string]*Schema
```

Comment:
Hmm, maybe it is as I see that callers don't have a `context.Context`.

Reply:
Usually I'd be all about it, but we would need to verify it can be properly passed from RPC logic all the way through to calling these functions. I'm curious about the expected need for it though? e.g. Is there a need to log anything in schema definitions?

Most provider code that I've run across with "schema functions" typically use function signatures without it. For some examples of what I mean:

- https://github.com/hashicorp/terraform-provider-aws/blob/4d0bdf093f90dc731c71b9c115c8bea2546c5ce2/internal/service/quicksight/schema/visual.go#L18
- https://github.com/hashicorp/terraform-provider-azurerm/blob/cd36aef1a2f87c1858fc1e0c08558cf2fb2f39e6/internal/services/compute/linux_virtual_machine_resource.go#L86
- https://github.com/cloudflare/terraform-provider-cloudflare/blob/553ce9fe3b3aa78677258d2af451d0b825ad2a16/internal/sdkv2provider/schema_cloudflare_record.go#L12

So adding the context will then require logic updates in those to pass the upstream context through all downstream schema function calls.

If having a proper context available is important for a use case, I'd suggest migrating the resource to the terraform-plugin-framework instead, which does already thread the context through properly, rather than spending additional time/effort on updating sdk-based code.

Reply:
No, I had no concrete use case in mind and there are other "Func"s that don't have a `Context` argument (e.g. `schema.StateFunc`), so please ignore my ramblings 😄.

Reply:
I took a quick peek at what it might take to support `context.Context` properly through here and unfortunately the internals meander into a lot of various functions/methods without it, including quite a few exported methods that would require breaking changes. While it would generally be possible to create duplicate, context-aware counterparts, I think its a whole lot of lift.